seerep_srv/seerep_core/src/core_dataset.cpp

```diff
@@ -888,37 +906,96 @@ void CoreDataset::intersectionDegree(const seerep_core_msgs::AABB& aabb,
-    xy_bounded = !(polygon_cgal.bounded_side(*vi) == CGAL::ON_UNBOUNDED_SIDE);
+    // is the vertex of the aabb inside or on the boundary (not outside) the
+    // polygon on the x and y axis
+    bool xy_bounded = !(cgal2.bounded_side(*vi) == CGAL::ON_UNBOUNDED_SIDE);
 
-    // check if this point is not inside the obb
-    if (!xy_bounded || !z_bounded)
+    if (!xy_bounded && checkIfFullyEncapsulated)
     {
       fullEncapsulation = false;
     }
-    else
+
+    if (xy_bounded && z_partially)
     {
       partialEncapsulation = true;
     }
   }
 }
 
+void CoreDataset::intersectionDegreeAABBinPolygon(
+    const seerep_core_msgs::AABB& aabb,
+    const seerep_core_msgs::Polygon2D& polygon,
+    CGAL::Polygon_2<Kernel> aabb_cgal, CGAL::Polygon_2<Kernel> polygon_cgal,
+    bool& fullEncapsulation, bool& partialEncapsulation)
+{
+  fullEncapsulation = true;
+  partialEncapsulation = false;
+
+  // are the corners of the aabb inside the polygon on the z axis
+  bool z_aabb_fully_in_polygon =
+      bg::get<bg::min_corner, 2>(aabb) >= polygon.z &&
+      bg::get<bg::min_corner, 2>(aabb) <= (polygon.z + polygon.height) &&
+      bg::get<bg::max_corner, 2>(aabb) >= polygon.z &&
+      bg::get<bg::max_corner, 2>(aabb) <= (polygon.z + polygon.height);
+  bool z_aabb_partially_in_polygon =
+      (bg::get<bg::min_corner, 2>(aabb) >= polygon.z &&
+       bg::get<bg::min_corner, 2>(aabb) <= (polygon.z + polygon.height)) ||
+      (bg::get<bg::max_corner, 2>(aabb) >= polygon.z &&
+       bg::get<bg::max_corner, 2>(aabb) <= (polygon.z + polygon.height));
+  // are the corners of the polygon inside the aabb on the z axis
+  /*bool z_polygon_fully_in_aabb =
+      polygon.z >= bg::get<bg::min_corner, 2>(aabb) &&
+      polygon.z <= bg::get<bg::max_corner, 2>(aabb) &&
+      (polygon.z + polygon.height) >= bg::get<bg::min_corner, 2>(aabb) &&
+      (polygon.z + polygon.height) <= bg::get<bg::max_corner, 2>(aabb);*/
+  bool z_polygon_partially_in_aabb =
+      (polygon.z >= bg::get<bg::min_corner, 2>(aabb) &&
+       polygon.z <= bg::get<bg::max_corner, 2>(aabb)) ||
+      ((polygon.z + polygon.height) >= bg::get<bg::min_corner, 2>(aabb) &&
+       (polygon.z + polygon.height) <= bg::get<bg::max_corner, 2>(aabb));
+
+  bool z_fully = z_aabb_fully_in_polygon;
+  bool z_partially = z_aabb_partially_in_polygon || z_polygon_partially_in_aabb;
+
+  // if not fully in z direction then for sure not fully
+  if (!z_fully)
+  {
+    fullEncapsulation = false;
+  }
+
+  // if not fully nor partially skip following checks
+  if (!z_fully && !z_partially)
+  {
+    return;
+  }
+
+  // traverse the axis aligned bounding box and check if AABB is in polygon
+  intersectionDegreeCgalPolygons(aabb_cgal, polygon_cgal, z_partially, true,
+                                 fullEncapsulation, partialEncapsulation);
+
+  // traverse polygon and check if it is fully in AABB
+  // if this is the check for AABB in polygon fail as all corners off AABB are
+  // outside of polygon
```

Comment:
```suggestion
  // traverse polygon and check if at least one vertex is in the AABB
```
Simplifying and fitting the explanation to the function call.